File: src/advent_of_code/2022/day02.clj
(ns advent-of-code.2022.day02
  "https://adventofcode.com/2022/day/2"
  (:require
   [clojure.string :as str]))

;;;; PART 1

;; possible moves
;; other player's move: A (Rock), B (Paper), C (Scissors)
;; my moves:            X (Rock), Y (Paper), Z (Scissors)

;; just 3x3 = 9 possibilities so I could just hard code them, and that might even be the best for readability

;; HOWEVER, I want to see if I can use some ascii math here
;; we can use 0 1 2 (lets call them `move-int`s) to represent rock paper and scissors respectively (subtracting from \A or \X)
;; then shape-score = (inc my-move-int)
;; for outcome-score, 
;;    (score 6): to win, my move-int has to be one ahead of other player's move (considering it as a cycle)
;;    (score 3): draw is easy both move-ints are the same
;;    (score 0): the only other case is loss (my move-int is two steps ahead/one step behind)
;;    lets say outcome-int = 0 1 2 for loss, draw and win (so that outcome-score = outcome-int * 3)

(def input-filename "inputs/2022/day02.txt")

(def rounds (->> (slurp input-filename)
                 (str/split-lines)))

(defn first-char-offset-from [s s-base]
  (let [char      (.charAt s 0)
        char-base (.charAt s-base 0)]
    (- (int char) (int char-base))))

(defn shape-score-fn [my-move-int]
  (inc my-move-int))

(defn outcome-int-fn [their-move-int my-move-int]
  ;; 0=loss 1=draw 2=win
  (-> (- my-move-int their-move-int)
      inc
      (mod 3)))

(defn outcome-score-fn [outcome-int]
  (* outcome-int 3))

(defn round-str->part-1-score [round-str]
  (let [[their-move my-move] (str/split round-str #" ")
        their-move-int (first-char-offset-from their-move "A")
        my-move-int    (first-char-offset-from my-move "X")
        shape-score    (shape-score-fn my-move-int)
        outcome-int    (outcome-int-fn their-move-int my-move-int)
        outcome-score  (outcome-score-fn outcome-int)]
    (+ shape-score outcome-score)))

(def part-1-answer
  "Total score I get if I follow the input plan"
  (->> rounds
       (map round-str->part-1-score)
       (apply +))
  )
;; 17189


;;;; PART 2

;; oooh two is interesting
;; A B C have the same definition

;; however, now X = Lose, Y = Draw, Z = win
;; 0 1 2 (outcome-int-fn) now corresponds to char - \X
;; 0 3 6 (scores)

;; so the problem is that I have to find my move given the other's move and the outcome

(defn my-move-int-fn [their-move-int outcome-int]
  (-> (+ their-move-int outcome-int)
      dec
      (mod 3)))

;; there must be a better way of doing these modulo arithmetic 
;; and by that I mean, given an algebraic equation with modulo, transfer them to other size so as to get eqn for another var
;; pretty sure I learnt this in school at some point, but I don't remember + don't have a piece of paper handy either
;; so I'm going to make sure this function works by doing a test across all 

;; making sure my-move-int fn defined above gives correct answer for all cases
;; check with outcome-int which I'm sure works because it gave correct answer in step 1
(assert
 (every? true? (for [their-move-i (range 3)
                     my-move-i    (range 3)]
                 ;; all potential (9) cases
                 (let [outcome-int (outcome-int-fn their-move-i my-move-i)]
                   ;; checking if doing my-move-int-fn and outcome-int-fn are complementary (is that the right word?)
                   (= my-move-i (my-move-int-fn their-move-i outcome-int)))))
 "my-move-int fn does not match up with outcome-int fn. At least one is broken")

(defn round-str->part-2-score [round-str]
  (let [[their-move outcome] (str/split round-str #" ")
        their-move-int       (first-char-offset-from their-move "A")
        outcome-int          (first-char-offset-from outcome "X")
        my-move-int          (my-move-int-fn their-move-int outcome-int)
        shape-score          (shape-score-fn my-move-int) 
        outcome-score        (outcome-score-fn outcome-int)]
    (+ shape-score outcome-score)))

(def part-2-answer
  "Total score I get assuming second input is the outcome"
  (->> rounds
       (map round-str->part-2-score)
       (apply +)))
;; 13490

(defn -main []
  (println "part-1 answer:" part-1-answer)
  (println "part-2 answer:" part-2-answer))

(comment
  (-main)
  )

File: python/2022/day02.py
"""Day 2: Rock Paper Scissors - https://adventofcode.com/2022/day/2"""

INPUT_FILE = "inputs/2022/day02.txt"


def parse_input(filename: str) -> list[tuple[str, str]]:
    """Parse input into list of (their_move, my_move/outcome) tuples."""
    with open(filename) as f:
        return [tuple(line.split()) for line in f.read().strip().split("\n")]


def char_offset(char: str, base: str) -> int:
    """Get offset of character from base character."""
    return ord(char) - ord(base)


def shape_score(my_move_int: int) -> int:
    """Score for the shape chosen (1 for Rock, 2 for Paper, 3 for Scissors)."""
    return my_move_int + 1


def outcome_int(their_move_int: int, my_move_int: int) -> int:
    """Calculate outcome: 0=loss, 1=draw, 2=win."""
    return (my_move_int - their_move_int + 1) % 3


def outcome_score(outcome: int) -> int:
    """Score for outcome (0, 3, or 6)."""
    return outcome * 3


def my_move_int_from_outcome(their_move_int: int, outcome: int) -> int:
    """Calculate my move given their move and desired outcome."""
    return (their_move_int + outcome - 1) % 3


def round_score_part1(their_move: str, my_move: str) -> int:
    """Calculate score for a round in part 1."""
    their_move_int = char_offset(their_move, "A")
    my_move_int = char_offset(my_move, "X")

    return shape_score(my_move_int) + outcome_score(outcome_int(their_move_int, my_move_int))


def round_score_part2(their_move: str, outcome_char: str) -> int:
    """Calculate score for a round in part 2."""
    their_move_int = char_offset(their_move, "A")
    outcome = char_offset(outcome_char, "X")
    my_move_int = my_move_int_from_outcome(their_move_int, outcome)

    return shape_score(my_move_int) + outcome_score(outcome)


def part1(rounds: list[tuple[str, str]]) -> int:
    """Total score if following part 1 interpretation."""
    return sum(round_score_part1(their, mine) for their, mine in rounds)


def part2(rounds: list[tuple[str, str]]) -> int:
    """Total score if following part 2 interpretation."""
    return sum(round_score_part2(their, outcome) for their, outcome in rounds)


def main():
    rounds = parse_input(INPUT_FILE)

    answer1 = part1(rounds)
    answer2 = part2(rounds)

    print(f"Part 1: {answer1}")  # 17189
    print(f"Part 2: {answer2}")  # 13490

    return answer1, answer2


if __name__ == "__main__":
    main()
